File: DFS of Graph - GFG/dfs-of-graph.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution {
  public:
    // Function to return a list containing the DFS traversal of the graph.
    vector<int> dfsOfGraph(int V, vector<int> adj[]) {
        // Code here
        for(int i=0;i<V;i++) 
            reverse(adj[i].begin(),adj[i].end());
            
        stack<int> s;
        int vis[V]={0};
        vector<int> ans;
        s.push(0);
        while(!s.empty()) {
            int t=s.top(); s.pop();
            if(!vis[t]) ans.push_back(t);
            vis[t]=1;
            for(auto i:adj[t]) {
                if(!vis[i]) s.push(i);
            }
        }
        return ans;
    }
};

//{ Driver Code Starts.
int main() {
    int tc;
    cin >> tc;
    while (tc--) {
        int V, E;
        cin >> V >> E;

        vector<int> adj[V];

        for (int i = 0; i < E; i++) {
            int u, v;
            cin >> u >> v;
            adj[u].push_back(v);
            adj[v].push_back(u);
        }
        // string s1;
        // cin>>s1;
        Solution obj;
        vector<int> ans = obj.dfsOfGraph(V, adj);
        for (int i = 0; i < ans.size(); i++) {
            cout << ans[i] << " ";
        }
        cout << endl;
    }
    return 0;
}
// } Driver Code Ends

File: DFS of Graph - GFG/dfs-of-graph.py
#User function Template for python3
from collections import deque
class Solution:
    
    #Function to return a list containing the DFS traversal of the graph.
    def dfsOfGraph(self, V, adj):
        # code here
        vis = set()
        res = []
        
        def dfs(node):
            res.append(node)            
            vis.add(node)
            for i in adj[node]:
                if i not in vis:
                    dfs(i)
            
        for i in range(V):
            if i not in vis:
                vis.add(i)
                dfs(i)
                
        return res;

#{ 
 # Driver Code Starts

if __name__ == '__main__':
    T=int(input())
    while T>0:
        V,E=map(int,input().split())
        adj=[[] for i in range(V+1)]
        for i in range(E):
            u,v=map(int,input().split())
            adj[u].append(v)
            adj[v].append(u)
        ob=Solution()
        ans=ob.dfsOfGraph(V,adj)
        for i in range(len(ans)):
            print(ans[i],end=" ")
        print()
        T-=1
# } Driver Code Ends